### ISSUE TITLE
Token not completely supported

### ISSUE THREAD

**Issue by wallyqs** [nats-io/nats-pure.rb#46]
Right now the only way to use `token` is by putting them into the URI string:

```ruby
nats.connect("nats://secret@nats:4222")
```

But in case there are special characters, this will fail because of URI encoded issues.  Ideally should be able to set token just with an option:

```ruby
nats.connect("nats://secret@nats:4222", token: 'my/token')
```

**Comment by ripienaar** [comment]
@wallyqs I kind of need these and also the ability to set custom jwt+signature callbacks like in go.

What would you think of this before I start spending time on tests?

```
diff --git a/lib/nats/io/client.rb b/lib/nats/io/client.rb
index 557715e..372303d 100644
--- a/lib/nats/io/client.rb
+++ b/lib/nats/io/client.rb
@@ -182,6 +182,9 @@ module NATS
       @user_nkey_cb = nil
       @user_jwt_cb = nil
       @signature_cb = nil
+
+      # Tokens
+      @auth_token = nil
     end

     # Establishes a connection to NATS.
@@ -254,10 +257,15 @@ module NATS
       end

       # NKEYS
+      @signature_cb ||= opts[:signature_cb]
+      @user_jwt_cb ||= opts[:user_jwt_cb]
       @user_credentials ||= opts[:user_credentials]
       @nkeys_seed ||= opts[:nkeys_seed]
       setup_nkeys_connect if @user_credentials or @nkeys_seed

+      # Tokens, if set will take preference over the user@server uri token
+      @auth_token ||= opts[:auth_token]
+
       # Check for TLS usage
       @tls = @options[:tls]

@@ -1037,7 +1045,7 @@ module NATS
         else
           cs[:auth_token] = @uri.user
         end
-      when @user_credentials
+      when @user_jwt_cb && @@signature_cb
         nonce = @server_info[:nonce]
         cs[:jwt] = @user_jwt_cb.call
         cs[:sig] = @signature_cb.call(nonce)
@@ -1047,6 +1055,8 @@ module NATS
         cs[:sig] = @signature_cb.call(nonce)
       end

+      cs[:auth_token] = @auth_token if @auth_token
+
       if @server_info[:headers]
         cs[:headers] = @server_info[:headers]
         cs[:no_responders] = if @options[:no_responders] == false
```

**Comment by wallyqs** [comment]
That looks good to me, there are a couple of branches right now in the repo so feel free to send to either and I will make the release:

- `main` which added the JS support but has some breaking compatibilities with `subscribe` (returns a subscription type now instead of a sid)
- `v0` which is branched off the past release in case you need a release based on that one without the features on `main`

**Comment by ripienaar** [comment]
Thanks, I probably need to move to `main` for my use anyway, quite unfortunate the breaking change :)

**Comment by ripienaar** [comment]
https://github.com/nats-io/nats-pure.rb/pull/56

**Comment by wallyqs** [comment]
Fixed via #56
